### PR TITLE
Fix `touch` to allow changing timestamps on directories, remake from #11760

### PR DESCRIPTION
**Pull request by kik4444**
Based off of #11760 to be mergable without conflicts.

# Description

Fix for #11757.
The main issue in #11757 is I tried to copy the timestamp from one directory to another only to realize that did not work whereas the coreutils `^touch` had no problems. I thought `--reference` just did not work, but apparently the whole `touch` command could not work on directories because `OpenOptions::new().write(true).create(true).open(&item)` tries to create `touch`'s target in advance and then modify its timestamps. But if the target is a directory that already exists then this would fail even though the crate used for working with timestamps, `filetime`, already works on directories. 

# User-Facing Changes

I don't believe this should change any existing valid behaviors. It just changes a non-working behavior.

# Tests + Formatting

~~I only could not run `cargo test` because I get compilation errors on the latest main branch~~
All tests pass with `cargo test --features=sqlite`

# After Submitting

### CODE REVIEW

File: crates/nu-command/src/filesystem/touch.rs [hunk 1]
```diff
@@ -75,88 +74,58 @@ impl Command for Touch {
...
+        if let Some(reference) = reference {
+            let reference_path = Path::new(&reference.item);
+            if !reference_path.exists() {
+                return Err(ShellError::FileNotFoundCustom {
+                    msg: "Reference path not found".into(),
+                    span: reference.span,
+                });
             }
+
+            let metadata = reference_path.metadata().expect("should be a valid path"); // Should never fail as the path exists
```
Comment (fdncred) [inline crates/nu-command/src/filesystem/touch.rs:95]: These `expect`s are the only thing that gives me pause about this PR. I've seen so many complaints about things like this panicking even though they shouldn't normally. I'd rather have some better error handling with these. Let's see what others think.

Comment (kik4444) [reply crates/nu-command/src/filesystem/touch.rs:95]: I've tried `touch`ing a file when I don't have read perms. In that case the `.exists()` should return false so we wouldn't get to the `expect`.

Have there been cases where something seemingly infallible like that fails?

Comment (fdncred) [reply crates/nu-command/src/filesystem/touch.rs:95]: mostly broken files or folders in messed up file systems. it's really hard to repro.

**Comment by fdncred** [comment]
Thanks!

**Pull request merged**